flink-table/flink-table-code-splitter/src/main/java/org/apache/flink/table/codesplit/BlockStatementSplitter.java

```diff
@@ -0,0 +1,331 @@
+
+        private final String parameters;
+
+        private int counter = 0;
+
+        private int bscCounter = 0;
+
+        private BlockStatementVisitor(String code, String context, String parameters) {
+            this.tokenStream = new CommonTokenStream(new JavaLexer(CharStreams.fromString(code)));
+            this.rewriter = new TokenStreamRewriter(tokenStream);
+            this.context = context;
+            this.parameters = parameters;
+        }
+
+        private BlockStatementVisitor(
+                TokenStreamRewriter rewriter, String context, String parameters) {
+            this.tokenStream = null;
+            this.rewriter = rewriter;
+            this.context = context;
+            this.parameters = parameters;
+        }
+
+        @Override
+        public Void visitStatement(StatementContext ctx) {
+
+            if (ctx.getChildCount() == 0) {
+                return null;
+            }
+
+            if (ctx.WHILE() != null) {
+                for (StatementContext statement : ctx.statement()) {
+                    BlockStatementVisitor visitor =
+                            new BlockStatementVisitor(
+                                    rewriter, context + "_whileBody" + counter++, parameters);
+                    children.add(visitor);
+                    visitor.visitStatement(statement);
+                }
+            } else if (ctx.IF() != null) {
+                for (StatementContext sc : ctx.statement()) {
+                    BlockStatementVisitor visitor =
+                            new BlockStatementVisitor(
+                                    rewriter, context + "_ifBody" + counter++, parameters);
+                    children.add(visitor);
+                    visitor.visitStatement(sc);
+                }
+
+            } else if (ctx.ELSE() != null) {
+                System.out.println(CodeSplitUtil.getContextString(ctx));
+            } else {
+                if (ctx.block() != null) {
+                    for (BlockStatementContext bsc : ctx.block().blockStatement()) {
+                        if (bsc.statement() != null
+                                && (bsc.statement().IF() != null
+                                        || bsc.statement().ELSE() != null
+                                        || bsc.statement().WHILE() != null)) {
+
+                            String localContext =
+                                    String.format("%s_%d", this.context, bscCounter++);
+                            BlockStatementVisitor visitor =
+                                    new BlockStatementVisitor(rewriter, localContext, parameters);
+                            visitor.addBlocks(extractedSingleBlocks);
+
+                            extractedSingleBlocks.clear();
+                            children.add(visitor);
+                            visitor.visitStatement(bsc.statement());
+                        } else {
+                            if (shouldExtract(ctx)) {
```

Comment:
Not all statements are added to `extractedSingleBlocks`. In that case `extractedSingleBlocks` is not a continuous block of code.

In `rewrite` you replace everything from the first element of `extractedSingleBlocks` to the last element. I guess this implies that `extractedSingleBlocks` is continuous. Why can you do this?

Reply:
HI @tsreaper,
so what maybe was not fully clear is that `extractedSingleBlocks` is a continuous block of single line statements up to next `block` expression so up to next IF/ELSE/WHILE block.

Lets break down below example:
Having:
```
public void myFun1(int[] a, int[] b) throws RuntimeException {
        if (a[0] == 0) {
            a[11] = b[0];
            a[12] = b[0];
            
            if (a[2] == 0) {
                a[2] = 1;
            } else {
                a[2] = b[2];
            }

            a[13] = b[0];
            a[14] = b[0];
        } else {
            a[0] = b[0];
            a[1] = b[1];
            a[2] = b[2];
        }
 }
```

With current implementation will be rewrite to:
  ```
public void myFun1(int[] a, int[] b) throws RuntimeException {
  
	if (a[0] == 0) {
		myFun1_rewriteGroup1(a, b);
  
		myFun1_ifBody0(a, b);
	} else {
		myFun1_ifBody1(a, b);
	}
  }
  
  void myFun1_rewriteGroup1(int[] a, int[] b) throws RuntimeException {
	myFun1_ifBody0_0(a, b);
	if (a[2] == 0) {
		a[2] = 1;
	} else {
		a[2] = b[2];
	}
  }
  
  void myFun1_ifBody0_0(int[] a, int[] b) throws RuntimeException {
	a[11] = b[0];
	a[12] = b[0];
  }
  
  void myFun1_ifBody1(int[] a, int[] b) throws RuntimeException {
	a[0] = b[0];
	a[1] = b[1];
	a[2] = b[2];
  }
  
  void myFun1_ifBody0(int[] a, int[] b) throws RuntimeException {
	a[13] = b[0];
	a[14] = b[0];
  }
```

The provided example has "non continuous block" of single line statements in `TRUE` branch of first level IF/ELSE block. The block is spitted by nested, 2nd level IF/ELSE block.

The first half of the single line statement block (a[11], a[12]) were grouped and extracted to new method together with 2nd level if/esle block - `myFun1_rewriteGroup1()`.In there we can see that those single line statements were extracted further to new method -> `myFun1_ifBody0_0`

The second half, expression a[13] and a[14] were extracted to their own method -> `myFun1_ifBody0`

While processing a statements, for example `TRUE` branch, whenever we will spot a new IF or ELSE we will "copy all single line statements" to newly created Visitor. The `extractedSingleBlocks` array for the "original visitor" will be cleared and ready to accept new single line statements. 

This is exactly what happen with a[11], a[12] from example above. They were added to the new Visitor created for 2nd level IF/ELSE. The a[13] and a[14] expressions were were added to cleared `extractedSingleBlocks` array in original Visitor starting a new continuous block.

Also maybe refer to this comment -> where I explained the high level logic https://github.com/apache/flink/pull/21393#discussion_r1068750626